# Metaheuristieken

## Optimalisatieproblemen (OPP)
Een optimalisatieprobleem (**OPP**) bestaat uit mogelijke **oplossingen**, een **doelfunctie** of een **kostenfunctie** en een lijst van beperkingen, **constraints**.
Een **OPP** omvat in de praktijk een aantal **variabelen**. Een oplossingen kan men bekomen door aan de variabelen een concrete waarde toe te kennen.
De constraints van een **OPP** bepaalt de **oplossingsruimte**. De oplossingsruimte bevat alle mogelijke oplossingen waaruit de besluiter kan kiezen.
De doelfunctie levert voor elke oplossingen een waarde op die aangeeft hoe goed de oplossing is.

## Computationele complexiteit (CC)
De computationele complexiteit van een probleem is een maat voor de hoeveelheid rekentijd en geheugen die nodig is om het probleem op te lossen.
Voor de **computationele complexiteit van een algortime** wordt meestal uit gegaan van de worst-case scenario.
Voor de **computationele complexiteit van een probleem** wordt uiteindelijk gekozen voor het meest efficiënte algoritme.
De **computationele complexiteit** van een probleem wordt vaak uitgedrukt in de **grootte van de invoer**. In de computationele complexiteit wordt ≈ aangeduid door de notatie **$O()$**.

Voorbeelden van computationele tijdscomplexiteit:

| Tijd                 | Notatie    | Voorbeeld                                     |
|----------------------|------------|-----------------------------------------------|
| Constant             | O(1)       | Het opvragen van een element in een array     |
| Lineair              | O(n)       | Het minimum in een ongeordende array vinden   |
| Logaritmisch         | O(log n)   | Binary search                                 |
| Lineair-logaritmisch | O(n log n) | Mergesort                                     |
| Kwadratisch          | O(n^2)     | Bubble sort, Selection sort, Quick sort       |
| Expontieel           | O(2^n)     | Recursief berekenen van het n-fibonacci getal |


### Polynomiale tijd
Een algoritme wordt in **polynomiale tijd** uitgevoerd als de computationele complexiteit begrensd is door een polynomiale functie van de grootte van de invoer.

In de computationele complexiteit beschouwt men meerdere complexiteitsklassen:
- **P**: Problemen die in polynomiale tijd opgelost kunnen worden door een deterministische Turingmachine.
- **NP**: Problemen waarvan de oplossing in polynomiale tijd kunnen worden opgelost door een niet-deterministische Turingmachine.
- **NP-complete**: dit is een subklasse van de klasse NP. Problemen die tot deze klasse behoren kunnen in polynomiale tijd op een deterministische Turingmachine omgezet worden naar een ander probleem in deze subklasse. Met ander woorden, indien voor één van de problemen behorende tot de klasse NP-complete een algoritme kan gevonden worden dat in polynomiale tijd op een deterministische Turingmachine kan worden opgelost, behoren in één klap alle problemen in de klasse NP-complete tot de klasse P.
- **NP-hard**: Alle problemen die niet behoren tot P of NP-complete.

## Waarom heuristieken?
Heuristieken worden gebruikt voor het oplossen van problemen waarvoor geen efficiënte algoritmes bestaan. De heuristieken zullen een goed antwoord geven in een veel kortere tijd dan het algoritme dat een perfect antwoord zal geven. een voorbeeld is het travelling salesman probleem. *Een handelsreiziger dient $n$ steden 1 maal te bezoeken en terug te keren naar zijn vertrekpunt. De handelsreiziger wil de kortst mogelijke route afleggen.* In totaal zijn er n! aantal mogelijke routes.
- Een algoritme om deze route optimale route te berekenen heeft een tijdscomplexiteit van $O(n² * 2^n)$
- Een heuristiek zoals de **nearest neighbour** heeft een tijdscomplexiteit van $O(n²)$ (De nearest neighbour is een algoritme dat de dichtsbijzijnde, niet bezochte, stad kiest en deze toevoegt aan de route)
Als we deze complexiteiten gebruiken en waarde geven aan $n$ kunnen we zien dat de heuristiek veel sneller is dan het algoritme.

| $n$ | $n!$                      | $n² * 2^n$  | $n²$ |
|-----|---------------------------|-------------|------|
| 5   | 120                       | 800         | 25   |
| 10  | 3.628.800                 | 102.400     | 100  |
| 15  | 1.307.674.368.000         | 7.372.800   | 225  |
| 20  | 2.432.902.008.176.640.000 | 419.430.400 | 400  | 

### Soorten heuristieken

#### 'Custom-made' heuristieken
Vele heuristieken zijn specifiek ontworpen voor een bepaald probleem. Deze kunnen niet gebruikt worden voor andere problemen.

#### Eenvoudige heuristieken
Eenvoudige heuristieken zijn vaak gebaseerd op een bepaalde strategie. Deze heuristieken zijn vaak niet optimaal, maar kunnen wel een goede oplossing geven.
Het *nearest neighbour* algoritme is een voorbeeld van een eenvoudige heuristiek.

##### 'constructie'-heuristieken
Een constructie-heuristiek bouwt een oplossing op door een aantal stappen te volgen. Een voorbeeld hiervan is het *nearest neighbour* algoritme.

##### 'twee-fasen'-heuristieken
Een twee-fasen-heuristiek bestaat uit twee stappen. De eerste stap is een constructie-heuristiek en de tweede stap is een verbeteringsheuristiek.

#### Meta-heuristieken
Een meta-heuristiek is een high level procedure dat op bijna elk probleem kan worden toegepast.

##### Simulated Annealing
Simulated Annealing is een meta-heuristiek gebaseerd op het proces van het afkoelen van metalen. Het algoritme zal een oplossing accepteren die slechter is dan de huidige oplossing. Dit wordt gedaan om uit het lokale minimum te geraken. Het algoritme zal na verloop van tijd minder slechtere oplossingen accepteren.

###### Python
Voor het uitvoeren van Simulated Annealing in python moet je eerst een subklasse aanmaken van de klasse Annealer waarin de functies **move** en **energy** worden gedefinieerd.
De functie **move** bepaalt hoe men van een oplossing naar een buuroplossing kan gaan.
De functie **energy** berekent de waarde van de objectieve functie van een oplossing.

```python
class Problem(Annealer):
    def move(self):
        pass

    def energy(self):
        pass
```

Als voorbeeld maken we gebruik van de **Rastrigin functie**. Voor de functie kiezen we 30 variabelen die in het interval [-5.12, 5.12] kunnen liggen.
De objectieve functie is gekend en ziet er als volgt uit:

$f(x) = 10*n+∑_{i=1}^{n}(x²-10*cos(2*π*x_{i}))$

Deze functie kunnen we schrijven in onze functie **energy**:

```python
def energy(self):
    sum_f = 10 * len(self.state)
    for i in range(0, len(self.state)):
        sum_f = sum_f + (self.state[i] ** 2 - 10 * math.cos(2 * math.pi * self.state[i]))
    return sum_f
```

We kunnen een initiële startoplossing, bestaande uit 30 random gegenereerde getallen tussen de gestelde onder- en bovengrens, genereren.

In [73]:
import numpy as np

init_sol = np.random.uniform(-5.12, 5.12, size=30)

Voor de overgang van een oplossing naar een buuroplossing kunnen we willekeurig een van de 30 dimensies selecteren en voor die dimensie een nieuwe waarde genereren.

```python
def move(self):
    i = np.random.randint(0, len(self.state))
    self.state[i] = np.random.uniform(-5.12, 5.12)
```

In [74]:
import math
from simanneal import Annealer

class RastriginProblem(Annealer):
    def move(self):
        i = np.random.randint(0, len(self.state))
        self.state[i] = np.random.uniform(-5.12, 5.12)
        
        
    def energy(self):
        sum_f = 10 * len(self.state)
        for i in range(0, len(self.state)):
            sum_f = sum_f + (self.state[i] ** 2 - 10 * math.cos(2 * math.pi * self.state[i]))
        return sum_f

Nu kunnen we een nieuw object aanmaken van de nieuwe klasse en de initiële oplossing meegeven

In [75]:
rastrigin = RastriginProblem(init_sol)
init_sol

array([ 2.39737412, -1.56868991,  1.95311905, -0.29843033, -3.43105703,
       -0.96364854,  5.01801604, -4.98539217, -1.89162231, -2.85355847,
       -1.43523489, -4.72062249,  3.27569475,  1.02003694,  2.92018899,
       -5.11873817,  0.40999543, -1.81885929,  0.57543031,  3.10712323,
        4.33159547, -2.60930614,  1.2946745 ,  3.88223591, -2.19931468,
        1.18597717,  4.70727975,  2.11316976, -1.59768719, -0.65591622])

Voor het te bekomen van een resultaat moeten we het simulated annealing proces laten uitvoeren

In [76]:
rastrigin.anneal()

 Temperature        Energy    Accept   Improve     Elapsed   Remaining
    2.50000         74.94    11.80%     6.00%     0:00:02     0:00:00

(array([-0.98435809, -0.92643434, -0.11957337, -1.9863871 , -1.00808321,
        -1.02601599,  1.04573653, -0.00977794, -0.12658816, -0.07554596,
         0.02504202, -1.04967754, -0.05500252,  0.0319085 , -0.03304844,
        -1.10620831, -2.03530224,  0.83135168,  1.99771853, -2.00040773,
         0.87239328, -0.01693468, -0.03542966, -0.09157199,  2.00149911,
         0.06195938, -0.07895591, -1.03340674,  0.03335864, -1.00561587]),
 55.973514061311974)

Met bovenstaande functie hebben we geen parameters meegeven dus zijn de default waarde gebruikt.
- Tmax = 25000.00 (de start temperatuur)
- Tmin = 2.5 (de stop temperatuur)
- temparature steps = 50.000 (het aantal iteraties)
- updates = 100 (het aantal updates)

We kunnen deze parameters ook zelf een waarde geven: